File: leetcode/841/841.go
package main

import "fmt"

func main() {
	// rooms := [][]int{{1}, {2}, {3}, {}, {4}} // false
	rooms := [][]int{{1}, {2}, {3}, {}} // true
	// rooms := [][]int{{1, 3}, {3, 0, 1}, {2}, {0}} // false
	// rooms := [][]int{{2, 3}, {}, {2}, {1, 3, 1}} // true

	fmt.Println(canVisitAllRooms(rooms))
}

func canVisitAllRooms(rooms [][]int) bool {
	var roomsUnlocked = make([]bool, len(rooms), len(rooms))
	roomsUnlocked[0] = true // room 0 starts unlocked

	numUnlocked := canVisitAllRoomsRec(rooms, roomsUnlocked, 0, 1)

	if numUnlocked == len(rooms) {
		return true
	}
	return false
}

func canVisitAllRoomsRec(rooms [][]int, roomsUnlocked []bool, roomNum int, numUnlocked int) int {

	// fmt.Println("roomNum:", roomNum)

	for _, key := range rooms[roomNum] {
		if roomsUnlocked[key] {
			continue
		}

		roomsUnlocked[key] = true

		// fmt.Println("roomsUnlocked:", roomsUnlocked)

		numUnlocked++

		numUnlocked = canVisitAllRoomsRec(rooms, roomsUnlocked, key, numUnlocked)
	}

	return numUnlocked

	// fmt.Println("numUnlocked:", numUnlocked)
}

File: leetcode/841/841.cpp
#include <iostream>
#include <vector>


class Solution {
public:
	int canVisitAllRoomsRec(std::vector<std::vector<int>>& rooms, std::vector<bool>& roomsUnlocked, int roomNum, int numUnlocked) {
		for (auto& key : rooms[roomNum]) {
			if (roomsUnlocked[key]) {
				continue;
			}

			roomsUnlocked[key] = true;

			numUnlocked++;

			numUnlocked = canVisitAllRoomsRec(rooms, roomsUnlocked, key, numUnlocked);
		}

		return numUnlocked;
	}

    bool canVisitAllRooms(std::vector<std::vector<int>>& rooms) {
		std::vector<bool> roomsUnlocked (rooms.size(), false);
		roomsUnlocked[0] = true; // room 0 starts unlocked

		int numUnlocked = canVisitAllRoomsRec(rooms, roomsUnlocked, 0, 1);

		if (numUnlocked == rooms.size()) {
			return true;
		}

        return false;
    }
};

int main() {
	Solution s;
	std::vector<std::vector<int>> rooms {{1}, {2}, {3}, {}};
	std::cout << s.canVisitAllRooms(rooms) << std::endl;
	return 0;
}